servlet/src/main/scala/org/http4s/servlet/ServletIo.scala

```diff
@@ -114,15 +113,18 @@ case class NonBlockingServletIo(chunkSize: Int) extends ServletIo {
           val blocked = Blocked(cb)
           state.getAndSet(blocked) match {
             case Ready if in.isReady =>
-              if (state.compareAndSet(blocked, Ready))
-                cb(read)
+              if (state.compareAndSet(blocked, Ready)) read(cb)
+
+            case Ready =>
+            /* NOOP: we wait for read, the `in.isReady` should trigger the listener */
+
             case Complete =>
-              if (state.compareAndSet(blocked, Complete))
-                cb(LeftEnd)
+              if (state.compareAndSet(blocked, Complete)) cb(LeftEnd)
+              else sys.error("Shouldn't get here.")
```

Comment:
Presumably we will never find out, but in the case we do I believe it will propagate to the servlet io thread. ```cb((new IllegalStateException(..)).left)``` would work as well but I really want this event to be as vocal as possible since it should never happen.